packages/core/src/utils.ts

```diff
@@ -259,3 +259,64 @@ export function errorAsyncIterator<TInput>(
 
   return stream;
 }
+
+export interface BasicCache<V> {
+  get(key: string): V | undefined;
+  set(key: string, value: V): void;
+  delete(key: string): void;
+  clear?: () => void;
+}
+
+export function getInMemoryLRUCache<V>({
+  max = 1000,
+  onDelete = () => {},
+}: {
+  max?: number;
+  onDelete?: (key: string, value: V) => void;
+} = {}): BasicCache<V> & { clear(): void } {
+  let storage: Record<string, V> = {};
+  let keys: string[] = [];
+  return {
+    get(key: string): V | undefined {
+      return storage[key];
+    },
+    set(key: string, value: V) {
+      queueMicrotask(() => {
+        if (storage[key] != null) {
+          if (keys.indexOf(key)) {
+            keys.splice(keys.indexOf(key), 1);
+          }
+        }
+        keys.push(key);
+        storage[key] = value;
+        while (keys.length > max) {
+          const key = keys.shift();
+          if (key != null) {
+            onDelete(key, storage[key]);
+            delete storage[key];
+          }
+        }
+      });
+    },
+    delete(key: string) {
+      queueMicrotask(() => {
+        if (storage[key] != null) {
+          keys.splice(keys.indexOf(key), 1);
+          onDelete(key, storage[key]);
+          delete storage[key];
+        }
+      });
+    },
+    clear() {
+      queueMicrotask(() => {
+        storage = {};
+        keys = [];
+      });
+    },
+  };
+}
```

Comment:
I am unsure whether this should be within core 🤔 
IMHO this could be a library outside of envelop as it is generic enough. I see that it is used within multiple envelop plugins, but not directly within the core.

Reply:
Yeah this can live as external lib (can also be a live under this repo i guess, at least for now)